### ISSUE TITLE
DOS programs can remember old network path

### ISSUE THREAD

**Issue by bolle732** [stsp/dosemu2#710]
Maybe this is not a bug of DOSEMU2! It's only for the record.

I had the strange effect with a copied installation of PC/GEOS.

When I started DOSEMU2, `lredir2.com` showed:
```
C:\ensemble>lredir2
Current Drive Redirections:
C: = \\LINUX\FS\home\<USER>\Programme\bbe413en-fdpp\.\drive_c\ attrib = READ/WRITE
cwd: /home/<USER>/Programme/bbe413en-fdpp/./drive_c/ENSEMBLE
```

Then I started PC/GEOS and found another drive "B:". The real problem was that it accessed the drive "B:" as the system drive instead if "C:"...

After existing PC/GEOS, `lredir2.com` showed:
```
C:\ensemble>lredir2
Current Drive Redirections:
B: = \\LINUX\FS\home\<USER>\Programme\bbe413en\.\drive_c\ attrib = READ/WRITE
C: = \\LINUX\FS\home\<USER>\Programme\bbe413en-fdpp\.\drive_c\ attrib = READ/WRITE
cwd: /home/<USER>/Programme/bbe413en-fdpp/./drive_c/ENSEMBLE
```
We got the drive "B:" from the original installation.

The solution with PC/GEOS is to reset its state by executing `reset.bat` or delete the files under `privdata\state`.

Don't know if other programs could go into the same problem. But I never saw it before, that PC/GEOS remembered old network drives and mount them automatically.

**Comment by bolle732** [comment]
So, I close this...

**Comment by stsp** [comment]
I dont think the program can remember
anything. It is much more likely that you
called lredir in some script.
Andrew, do you know if the redirection
can be duplicated with subst? I dont think
so but havent checked.

**Comment by bolle732** [comment]
I've double/triple checked everything, there is no location where I use lredir. And the output shown above is before starting PC/GEOS and after exiting it.

I'm sure that PC/GEOS has triggered somehow the lredir. After removing the state files of PC/GEOS, everything was ok.

It's a mystery to myself.

**Comment by stsp** [comment]
If you can provide the geos files that do
this (I already have the basic installation),
then it would be interesting to take a look.

**Comment by andrewbird** [comment]
> Andrew, do you know if the redirection
can be duplicated with subst? I dont think
so but havent checked.

SUBST would be purely on a DOS level, whereas here both CDS (as indicated by lredir not showing DISABLED attribute) and lredir (drives[].root must be set) believe the drive is redirected.
I might have thought that the init process redirecting drives had somehow globbed an old directory, but the fact that the drive is `B:` pretty much rules that one out.
As Stas says, a stray call to lredir is most likely the culprit, but the interface that lredir uses to create drives is standard DOS, so perhaps geos calls int21/5f03 itself to recreate the mapping.
It would be really interesting to see the boot.log (with debug -D+d).

**Comment by bolle732** [comment]
> If you can provide the geos files that do
> this (I already have the basic installation),
> then it would be interesting to take a look.

I'll see if I can replicate the problem first. Hopefully I can do it today.

**Comment by andrewbird** [comment]
So I know very little about geos, but can you map / unmap drives using its filemanager etc? If so try mapping say `E`: to `\\LINUX\FS\tmp` then close down geos and restart to see if the drive persists? If it does, I guess that's a feature :smile: If it doesn't we should keep looking.

**Comment by bolle732** [comment]
There was once a version (as IBM EduQuest aka SchoolView) which could map network drives (Netware shares).

But that it still has a functionality to map drives would be new to me.

**Comment by andrewbird** [comment]
Okay, I'm out of ideas, so let's see if you can recreate the problem and produce the boot.log with -D+d debug

**Comment by bolle732** [comment]
Ok, I can reproduce the problem. But now it has the old path on "A:"...

- Old custom setup in `~/Programme/bbe413en`
- Create a new custom setup under `~/Programme/bbe413en-fdpp`
- Copy the PC/GEOS folder `~/Programme/bbe413en/drive_c/ensemble` to a new custom setup ` ~/Programme/bbe413en-fdpp/drive_c` with `cp -a`

Deleting the state file `drive_c/ensemble/privdata/state/ui_state.000` solves the ghost lredir.

So, it seems really to be a feature ;-) Would it be possible to write a PC/GEOS programto map drives with the help of lredir ?

If still interested, I can provide a log file. Is `-D+d` enough ?

**Comment by bolle732** [comment]
I did the last check and copied only the offending state file `ui_state.000`. It restores the ghost lredir.

**Comment by andrewbird** [comment]
`-D+d` is enough for me, I think it should expose when the drive is mapped.

**Comment by bolle732** [comment]
Here it is:
[boot.log.gz](https://github.com/stsp/dosemu2/files/2561820/boot.log.gz)

**Comment by andrewbird** [comment]
It certainly looks like geos is mapping it. You can see the existing mapped drives being listed(just `C:`), then drive `B:` is added.

**Comment by stsp** [comment]
But how does it do dosemu2-specific unc?
Really strange.

**Comment by andrewbird** [comment]
I figure it remembered the output of the prior listing of mapped drives

**Comment by stsp** [comment]
How did it got the listing?

**Comment by andrewbird** [comment]
int 21/5f02

**Comment by bolle732** [comment]
So, this is a specific feature of PC/GEOS. Never mind it and always <funny> news...

**Comment by stsp** [comment]
> int 21/5f02

Is this seen in the log?
Interesting.
So does it see that the path of drive C have
changed from the prev invocation:
```
B: = \\LINUX\FS\home\<USER>\Programme\bbe413en\.\drive_c\ attrib = READ/WRITE
C: = \\LINUX\FS\home\<USER>\Programme\bbe413en-fdpp\.\drive_c\ attrib = READ/WRITE
```
(note the slight difference)
... and adds the previous C drive as B?

**Comment by andrewbird** [comment]
Yep, int21/5f02 is seen in the log, I mentioned that it lists the current redirections before adding the `B:` drive. I figure that it notices that `C:` is already mapped and picks a free one.

**Comment by stsp** [comment]
I wonder if this subtle difference in the UNC
path makes it to do so. I suppose be there the
identical dirs, it won't be?

**Comment by stsp** [comment]
@bolle732 so what do you think about this
subtle difference in 2 UNC pathes? Isn't this
a source of the problem?

**Comment by bolle732** [comment]
> @bolle732 so what do you think about this
subtle difference in 2 UNC pathes? Isn't this
a source of the problem?

Do you mean the difference of the base folder `bbe413en` vs `bbe413en-fdpp` ? I think a single different letter should be enough. But who knows, if a crazy programmer has implemented an heuristic... Maybe that could be a problem converting between LFN and SFN.

One information not present in the `boot.log` of the problematic computer is the special user home path:

`/home/<USER>/` -> `/home/<DOMAIN>.<TLD>/<USER>/`

This machine is joined into an LDAP and therefore an additional folder with the internal domain name exist in the path. Don't know if this could have some influence too.

I have two (ok, multiple) custom setups with a base installation of PC/GEOS. One using standard SNF and the other LFN. The problematic installation is using SFN, but I haven't tested with LFN. Changing an installation to LFN and vice versa is not trivial. One would normally do a fresh install.

Further due to messing around lately with my custom setups, my installation has diverged a little bit. 

I've tried to replicate the problem on another notebook with an installation using SFN. But on that it seems to work always fine. I couldn't reproduce it up to now :(

Will see that I can give it another try tomorrow when I have access to the original machine.

Last but not least, I'll ask the one owning all the source code of PC/GEOS. Maybe I could meet him in a few days on a GEOS reunion.

**Comment by andrewbird** [comment]
> Let me see if I can find out more at the GEOS meeting.

Do the authors of PC/GEOS still make money from selling it or services around it? I wonder if there's an appetite yet for releasing code as open source like MS did with DOS 1 & 2 or the PC-MOS people? Sometimes I feel like an archeologist when digging around for info on old software, and I do think it's becoming more difficult with sites disappearing. Whether people actually use or further develop the software is not really likely in the case of MS-DOS 1 / 2, and not really a concern, but it's more that unless companies do this sort of gesture that piece of history will be lost on some long forgotten backup or archive tape forever.

**Comment by bolle732** [comment]
If everything goes well, I could say more after the GEOS meeting. As far as I know, there are plans to make it open source, but I don't know under what type of license. The main problem here is, that some parts are licensed from 3rd parties. And after 30 years, it's often unknown, if and who held the current rights on this parts.

**Comment by stsp** [comment]
This would mean the first rather functional
dos-based os open sourced. Looking
forward for this to happen. :) Is there
some internet browsing btw?

**Comment by bolle732** [comment]
Regarding the future, there is actually not too much to be found. There have been some discussion on the forum of the German based website GEOS-Infobase. The announcement has been a year ago:

http://geos-infobase.de/WBB_317/index.php?page=Thread&threadID=2469&s=2e1a788c1de0c9f757d5a7fc6abfed557504e1f4

Right now, we still don't know much details. Especially as the new holder talks about open sourcing the code for the community, but also wrote they are planing some services on top of it.

On the next GEOS meeting which is held the next week, we should hear more definitive details:

https://www.moellerjaner.de/pc-geos-treffen-2018-in-syhra/

There, the new holder will talk about the future and gives a workshop regarding to build your own SDK for PC/GEOS. That will attract me ;-)

**Comment by stsp** [comment]
They mention dosemu there.
Is this dosemu2?

**Comment by bolle732** [comment]
If you're asking about the mention of dosemu in the first link, then it would be dosemu2.

In the forum, you'll find posts regarding dosemu1 and 2. Sorry, could be my fault to start there talking about the best environment for PC/GEOS...

Today, PC/GEOS Linux users may mostly using dosemu2 or the version provided in the repository of their distribution (1 or 2). Some may still have a 32-bit version with an old dosemu1.
I've did once a statically compiled dosemu1 version which I've installed at a meeting some years ago on about 10 machines.

Users with Windows / MacOSX uses DOSBOX.

And there are still the ones using it under native DOS.

**Comment by stsp** [comment]
I wish I had named the project differently.
Now 2 is just a version number, and the project is... dosemu.
FDPP is a much smarter choice: no "FreeDos" within, so
its recognizable, but "FD" letters still give the deserved credits
to FreeDOS ppl.
dosemu2 was a looser's naming. :)
Now I am doomed to ask "was this dosemu 1 or 2".

**Comment by bolle732** [comment]
The free parts of PC/GEOS the source code is now public:

https://github.com/bluewaysw

It still needs some fixes that the tool chain so that it can be compiled completely and for the missing parts, there has to be created a free replacement.

**Comment by stsp** [comment]
This is cool!
I suppose you already notified the
freedos ppl about this, hope they can
help with getting it compiled.

**Comment by andrewbird** [comment]
Great stuff, it'll make interesting reading, thanks!

**Comment by stsp** [comment]
If he wants to do some work beforehand,
we can't pre-announce the project.

**Comment by bolle732** [comment]
> Great stuff, it'll make interesting reading, thanks!

Yes, there are a lot of great concepts / ideas in PC/GEOS, which the products of some global players still lack about.

To get a higher level of overview, an interested person could / should read the Concept Books of the GEOS SDK. I haven't seen them right now on the repository, but they could be included in the downloads found at http://geos-sdk.com/.

To understand PC/GEOS, one has to invest some hours of reading. Some have invested days, weeks, months or years and still haven't understood. I know a little bit, have done some (simple) programming with it. But I'm still far away from the levels that others have reached...

**Comment by stsp** [comment]
I wouldn't bet on people to start coding
free software with geos sdk. :) DOS-based
development platform is not what everyone
is looking for.
If geos has a lot of interesting things that
already work, then that may be good for the
dos hobbyists.

**Comment by stsp** [comment]
> The free parts of PC/GEOS the source code is now public:

Its written on a 16bit asm...
Is this understanding correct?
32bit/prot mode apps are impossible there, is this correct?
Sigh.

**Comment by bolle732** [comment]
Yes, PC/GEOS was written in 16bit object orientated assembler, called Espire (ESP). Later, there came the SDK, where you can write either in ESP or GOC, where GOC is an OO extension to C like Apples Objective C.

There has been tests to move from 16bit real mode to 286 or 386 protected mode. But this has never been available before to the public. I can't say for now, if this parts are inside of the current source.

If you really want to know, you should ask Falk.

**Comment by stsp** [comment]
No, I'd be interested if there is a move to C, at
least in some kernel parts.
How did they suppose to do the android port?
You can't port 16bit asm to android, or can you?

**Comment by bolle732** [comment]
As I remember, yes, there was another effort of moving from most ASM code to C. But I never seen that released.

I don't know about a native Android version. If I'm correct, the last project from the old holder (Breadbox) has been using a DOSBox variant running PC/GEOS on an Android tablet.

**Comment by andrewbird** [comment]
It looks like word is getting around
http://www.vcfed.org/forum/showthread.php?67185-PC-GEOS-Source-code-amp-Apps-now-available-on-Git!

**Comment by bolle732** [comment]
Yes, that had to be expected :-)

Thanks for sharing the links!

**Comment by stsp** [comment]
This link says "Hello World".

**Comment by bolle732** [comment]
Yes, that's correct. You have to to use the "Register" link in the upper right corner. After registering, you should receive an email with the instruction to download.
